### PR TITLE
fix(sourcemaps): Add bundle_id injection and version support

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at adding support for `debug_id` of the bundle and also the possibility to upload a bundle with a `release` and/or `dist`.

### CODE REVIEW

File: src/sentry/tasks/assemble.py [hunk 1]
```diff
@@ -251,53 +251,68 @@ def _normalize_debug_id(debug_id: Optional[str]) -> Optional[str]:
         return None
 
 
-def _extract_debug_ids_from_manifest(manifest: dict) -> List[Tuple[SourceFileType, str]]:
+def _extract_debug_ids_from_manifest(
+    manifest: dict,
+) -> Tuple[str, List[Tuple[SourceFileType, str]]]:
```
Comment (Swatinem) [inline src/sentry/tasks/assemble.py:256]:
```suggestion
) -> Tuple[Optional[str], List[Tuple[SourceFileType, str]]]:
```

As the `bundle_id` could be missing as well?

File: src/sentry/tasks/assemble.py [hunk 2]
```diff
@@ -311,14 +326,94 @@ def _create_artifact_bundle(
...
+    if release and dist_name:
+        dist = release.add_dist(dist_name)
+
+    artifact_count = len(manifest.get("files", {}))
+    min_artifact_count = options.get("processing.release-archive-min-files")
+    saved_as_archive = False
+
+    if artifact_count >= min_artifact_count:
+        if release is None:
+            raise AssembleArtifactsError("release does not exist")
```
Comment (Swatinem) [inline src/sentry/tasks/assemble.py:371]: didn’t we already asserting this a couple of lines above?

File: src/sentry/tasks/assemble.py [hunk 3]
```diff
@@ -311,14 +326,94 @@ def _create_artifact_bundle(
...
+
+        try:
+            update_artifact_index(release, dist, bundle)
+            saved_as_archive = True
```
Comment (Swatinem) [inline src/sentry/tasks/assemble.py:375]: this means we store single files as a fallback if updating the index failed for whatever reason. is that intentional?

**Review by Swatinem — APPROVED** [review]
tbh, the logic is a bit hard to follow. But maybe its just that I don’t understand the difference between `artifact.bundle` and `release.bundle`? It is the same file format after all?

**Comment by iambriccardo** [comment]
> tbh, the logic is a bit hard to follow. But maybe its just that I don’t understand the difference between `artifact.bundle` and `release.bundle`? It is the same file format after all?

These changes were kept independent to avoid touching the core logic that was present before for the release bundles (which is a bit weird).

The difference between them is that new bundles are independent of releases and thus we name them artifact bundles instead of release bundles.

File: src/sentry/tasks/assemble.py (hunk 2 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:371]: Nope, this assertion is that the actual release is existing in the db. The above's assertion checks whether we have a manifest/request data mismatch.

File: src/sentry/tasks/assemble.py (hunk 1 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:256]: Thanks! Yes

File: src/sentry/tasks/assemble.py (hunk 3 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:375]: This was all the previous logic so I assume it is.

File: src/sentry/tasks/assemble.py [hunk 4]
```diff
@@ -311,14 +326,94 @@ def _create_artifact_bundle(
...
+        raise AssembleArtifactsError("organization does not match uploaded bundle")
+
+    release_name = manifest.get("release")
+    if release_name != version:
+        raise AssembleArtifactsError("release does not match uploaded bundle")
+
+    try:
+        release = Release.objects.get(organization_id=organization.id, version=release_name)
+    except Release.DoesNotExist:
+        raise AssembleArtifactsError("release does not exist")
```
Comment (Swatinem) [inline src/sentry/tasks/assemble.py:358]: But doesn’t this mean that you either have a `release`, or otherwise you throw?

In what circumstances can `release` still be `None` after this block?

Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:358]: I can confirm, I overlooked this part during the split of the functions, will fix it now.

File: src/sentry/tasks/assemble.py [hunk 5]
```diff
@@ -251,53 +251,68 @@ def _normalize_debug_id(debug_id: Optional[str]) -> Optional[str]:
...
+        bundle_id, debug_ids_with_types = _extract_debug_ids_from_manifest(archive.manifest)
 
-        if len(debug_ids_with_types) > 0:
+        # We want to save an artifact bundle only if we have found debug ids in the manifest or if the user specified
+        # a release for the upload.
+        if len(debug_ids_with_types) > 0 or version is not None:
             artifact_bundle = ArtifactBundle.objects.create(
                 organization_id=org_id,
-                bundle_id=uuid.uuid4().hex,
+                bundle_id=bundle_id,
```
Comment (mitsuhiko) [inline src/sentry/tasks/assemble.py:300]: I think for new uploads with should default the `bundle_id` here (eg: generate a random UUID if none is supplied). Logic: that way we can then start encoding the bundle_id/debug_id into the URL for deleting artifact bundles that are not release bound.

**Comment by mitsuhiko** [comment]
This seems fine but I think we should address the `bundle_id` aspect. I want to avoid the case where we have non release uploads without bundle id.

File: src/sentry/tasks/assemble.py [hunk 6]
```diff
@@ -251,53 +251,68 @@ def _normalize_debug_id(debug_id: Optional[str]) -> Optional[str]:
...
+    # ReleaseArtifactBundle table.
+    bundle_id = manifest.get("debug_id")
+    if bundle_id is not None:
+        bundle_id = _normalize_debug_id(bundle_id)
+
     files = manifest.get("files", {})
     for file_path, info in files.items():
         headers = _normalize_headers(info.get("headers", {}))
-        if (debug_id := headers.get("debug-id", None)) is not None:
+        if (debug_id := headers.get("debug-id")) is not None:
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py:271]: Is there a difference in python between `headers.get("debug-id", None) is not None:` and ``headers.get("debug-id") is not None:`? I've seen we provide default value for getters, and then use `is/is not` immediately after this in a few places (eg. in `allow_scraping` in processor)

File: src/sentry/tasks/assemble.py [hunk 7]
```diff
@@ -251,53 +251,68 @@ def _normalize_debug_id(debug_id: Optional[str]) -> Optional[str]:
...
             )
 
-            if release:
+            # If a release version is passed, we want to create the weak association between a bundle and a release.
+            if version is not None:
                 ReleaseArtifactBundle.objects.create(
                     organization_id=org_id,
-                    release_name=release.version if release else None,
-                    dist_name=dist.name if dist else None,
+                    release_name=version,
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py:309]: Why do we use `name` instead of `id` for both things here?

File: src/sentry/tasks/assemble.py [hunk 8]
```diff
@@ -311,14 +326,91 @@ def _create_artifact_bundle(
...
+        else:
+            raise AssembleArtifactsError(
+                "uploading a bundle without debug ids or release is prohibited"
+            )
+
+
+def get_with_manifest_fallback(value, manifest_key, manifest, throw=True):
+    if value is None:
+        manifest_value = manifest.get(manifest_key)
+        if manifest_value is None and throw:
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py:338]: Hmm, what happens if `dist` is correctly loaded, but it was `null` in manifest, thus return `None` here? Will it throw? Or do we not store `null` in the manifest and it will barf because of the missing key? `dist` is optional in the end.

File: src/sentry/tasks/assemble.py [hunk 9]
```diff
@@ -311,14 +326,91 @@ def _create_artifact_bundle(
...
+
+    try:
+        release = Release.objects.get(organization_id=organization.id, version=release_name)
+    except Release.DoesNotExist:
+        raise AssembleArtifactsError("release does not exist")
 
-            return True
+    dist_name = manifest.get("dist")
+    dist = None
+    if release and dist_name:
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py:362]: It's impossible for `release` to not be defined at this stage due to `except Release.DoesNotExist` so you can drop this check and write

```
dist = release.add_dist(dist_name) if dist_name else None
```

File: src/sentry/tasks/assemble.py [hunk 10]
```diff
@@ -311,14 +326,91 @@ def _create_artifact_bundle(
...
+    return value
+
+
+def handle_assemble_for_release_file(bundle, archive, manifest, organization, version):
+    org_slug = manifest.get("org")
+    if organization.slug != org_slug:
+        raise AssembleArtifactsError("organization does not match uploaded bundle")
+
+    release_name = manifest.get("release")
+    if release_name != version:
```
[the remark this answers is not in the document]
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:352]: I found this strange too, I just kept the old code in order to avoid regressions (maybe during renaming some problems would come up). All of this under the assumption that we are going to fade it out.

I can follow up with a PR to actually improve all the code but I am not sure it's worth it.

File: src/sentry/tasks/assemble.py (hunk 9 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:362]: Overlooked that during refactor, thanks!

File: src/sentry/tasks/assemble.py (hunk 8 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:338]: That's a good observation. In my mind I assumed that None set is equal to unset.

File: src/sentry/tasks/assemble.py [hunk 11]
```diff
@@ -311,14 +326,91 @@ def _create_artifact_bundle(
...
+        _store_single_files(archive, meta, True)
+
+    return artifact_count
+
+
+def handle_assemble_for_artifact_bundle(bundle, manifest, organization, version, project_ids, dist):
+    # We want to give precedence to the request fields and only if they are unset fallback to the manifest's
+    # contents.
+    release_name = get_with_manifest_fallback(version, "release", manifest, throw=False)
+    dist_name = get_with_manifest_fallback(dist, "dist", manifest, throw=False)
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py:391]: Why so overengineered? 😅

```suggestion
    release_name = version or manifest.get("release")
    dist_name = dist or manifest.get("dist")
```

File: src/sentry/tasks/assemble.py [hunk 12]
```diff
@@ -353,64 +446,26 @@ def assemble_artifacts(org_id, version, checksum, chunks, project_ids=None, **kw
...
-
-            # Count files extracted, to compare them to release files endpoint
-            metrics.incr("tasks.assemble.extracted_files", amount=artifact_count)
+            if upload_as_artifact_bundle:
+                artifact_count = handle_assemble_for_artifact_bundle(
+                    bundle, manifest, organization, version, project_ids, dist
+                )
+            else:
+                artifact_count = handle_assemble_for_release_file(
+                    bundle, archive, manifest, organization, version
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py:464]:
```suggestion
                    bundle, manifest, organization, version, archive
```

Maybe we can keep it consistent like this?

File: src/sentry/tasks/assemble.py (hunk 8 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:338]: With @mitsuhiko we agreed that we want to give priority to the values in the request instead of the manifest.

File: src/sentry/api/endpoints/organization_artifactbundle_assemble.py [hunk 13]
```diff
@@ -77,13 +80,25 @@ def post(self, request: Request, organization) -> Response:
 
         from sentry.tasks.assemble import assemble_artifacts
 
+        version = data.get("version")
+        dist = data.get("dist")
+
+        if version is None and dist is not None:
+            return Response(
+                {"error": "You need to specify a version together with a dist"}, status=400
```
[the remark this answers is not in the document]
Comment (iambriccardo) [reply src/sentry/api/endpoints/organization_artifactbundle_assemble.py:88]: I agree this is super confusing, I will take care of refactoring it across all the code. I was using `version` since it was already present in async job params. I will keep it as version there to avoid issues of old code calling with keyword version on the new code.

File: src/sentry/tasks/assemble.py (hunk 7 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:309]: We decided with @mitsuhiko that we want a weak reference to both, which means that we can bind a release/dist pair that have not been created in the db. The goal is to have in ci the ability to specify a release which will actually be created in a future step during deployment. Ofc this adds the complexity of having to keep the namings consistent.

File: src/sentry/tasks/assemble.py (hunk 6 above)
Comment (kamilogorek) [reply src/sentry/tasks/assemble.py:271]: That was my thinking, and that's why I did not understand why we had `get(key, def) is not def` in few places :P

File: src/sentry/tasks/assemble.py (hunk 11 above)
Comment (iambriccardo) [reply src/sentry/tasks/assemble.py:391]: I have actually used the `throw` in the `handle_assemble_for_release_file` method but then decided to keep the old logic since the previous implementation required consistent values across request/manifest instead of having the manifest as a fallback and I forgot to remove this convoluted function. Thanks!

File: src/sentry/tasks/assemble.py [hunk 14]
```diff
@@ -332,37 +333,26 @@ def _create_artifact_bundle(
...
     try:
-        release = Release.objects.get(organization_id=organization.id, version=release_name)
+        release = Release.objects.get(organization_id=organization.id, version=version)
     except Release.DoesNotExist:
         raise AssembleArtifactsError("release does not exist")
 
     dist_name = manifest.get("dist")
     dist = None
-    if release and dist_name:
+    if dist_name:
```
Comment (kamilogorek) [inline src/sentry/tasks/assemble.py]:
```
dist = release.add_dist(dist_name) if dist_name else None
```
